### PR TITLE
Add support for IndieAuth metadata endpoint

### PR DESCRIPTION
**Pull request by gRegorLove**
- Adds `discoverMetadataEndpoint()`: if a metadata endpoint is found, this fetches the JSON, stores it, stores a decoded PHP array version, and returns the metadata endpoint
    - Adds `setMetadata()`, `getMetadata()`, and `resetMetadata()` for this and for easier unit testing
- Adds `discoverRevocationEndpoint()`, `discoverIntrospectionEndpoint()`, and
`discoverUserinfoEndpoint()` methods
- If metadata endpoint is found, checks for `issuer` in the body and ensures it's valid with `_isIssuerValid()`; returns `invalid_issuer` error message if it isn't.
- `_discoverEndpoint()` updated to first check for endpoints in the parsed metadata with `_discoverFromMetadata()`, then continues to check in HTTP headers and rel links if not found
- `build_url()` updated to ensure scheme and hostname are lowercased for URL normalization
- Adds MetadataTest.php with unit tests for these changes
- Some minor syntax improvements, adding braces on one-line if conditions

### CODE REVIEW

**Review by dshanske — APPROVED** [review]
Looks good to me.

**Comment by dshanske** [comment]
@gRegorLove You didn't add in https://indieauth.spec.indieweb.org/#authorization-response verifying the issuer parameter. Separate PR?

**Comment by gRegorLove** [comment]
Thanks @dshanske, just pushed an update. For backcompat, I set it to only verify the `iss` if a metadata endpoint was found with `issuer` URL.

File: src/IndieAuth/Client.php [hunk 1]
```diff
@@ -113,6 +115,16 @@ public static function complete($params) {
       return self::_errorResponse('invalid_state', 'The authorization server returned an invalid state parameter');
     }
 
+    if (isset($_SESSION['indieauth_issuer'])) {
+      if (!isset($params['iss'])) {
+        return self::_errorResponse('missing_iss', 'The authorization server did not return the iss parameter');
+      }
+
+      if ($params['iss'] !== $_SESSION['indieauth_issuer']) {
+        return self::_errorResponse('invalid_iss', 'The authorization server return an invalid iss parameter');
```
Comment (dshanske) [inline src/IndieAuth/Client.php]: returned, not return.

**Comment by gRegorLove** [comment]
I need to make one more update then I think this will be ready. Needs to fetch metadata again before [this line](https://github.com/indieweb/indieauth-client-php/blob/main/src/IndieAuth/Client.php#L141) in `complete()`.

**Review by dshanske — APPROVED** [review]

**Review by martymcguire — APPROVED** [review]
The new changes seem good! I must admit I only skimmed the changes to the unit tests, but I checked out this branch and they ran fine for me!

**Pull request merged**